tests/components/bluesound/test_media_player.py

```diff
@@ -0,0 +1,348 @@
+    )
+
+    player_mocks.player_data.player.volume.assert_called_once_with(level=50)
+
+
+async def test_volume_mute(
+    hass: HomeAssistant, setup_config_entry: None, player_mocks: PlayerMocks
+) -> None:
+    """Test the media player volume mute."""
+    await hass.services.async_call(
+        "media_player",
+        "volume_mute",
+        {"entity_id": "media_player.player_name1111", "is_volume_muted": True},
+        blocking=True,
+    )
+
+    player_mocks.player_data.player.volume.assert_called_once_with(mute=True)
+
+
+async def test_volume_up(
+    hass: HomeAssistant, setup_config_entry: None, player_mocks: PlayerMocks
+) -> None:
+    """Test the media player volume up."""
+    await hass.services.async_call(
+        "media_player",
+        "volume_up",
+        {"entity_id": "media_player.player_name1111"},
+        blocking=True,
+    )
+
+    player_mocks.player_data.player.volume.assert_called_once_with(level=11)
+
+
+async def test_volume_down(
+    hass: HomeAssistant, setup_config_entry: None, player_mocks: PlayerMocks
+) -> None:
+    """Test the media player volume down."""
+    await hass.services.async_call(
+        "media_player",
+        "volume_down",
+        {"entity_id": "media_player.player_name1111"},
+        blocking=True,
+    )
+
+    player_mocks.player_data.player.volume.assert_called_once_with(level=9)
+
+
+async def test_attributes_set(
+    hass: HomeAssistant, setup_config_entry: None, player_mocks: PlayerMocks
+) -> None:
+    """Test the media player attributes set."""
+    state = hass.states.get("media_player.player_name1111")
+    assert state.state == "playing"
+    assert state.attributes["volume_level"] == 0.1
+    assert state.attributes["is_volume_muted"] is False
+    assert state.attributes["media_content_type"] == "music"
+    assert state.attributes["media_position"] == 2
+    assert state.attributes["shuffle"] is False
+    assert state.attributes["master"] is False
+    assert state.attributes["friendly_name"] == "player-name1111"
+    assert state.attributes["media_title"] == "song"
+    assert state.attributes["media_artist"] == "artist"
+    assert state.attributes["media_album_name"] == "album"
+
+
+async def test_status_updated(
+    hass: HomeAssistant,
+    setup_config_entry: None,
+    player_mocks: PlayerMocks,
+) -> None:
+    """Test the media player status updated."""
+    pre_state = hass.states.get("media_player.player_name1111")
+    assert pre_state.state == "playing"
+    assert pre_state.attributes["volume_level"] == 0.1
+
+    status = player_mocks.player_data.status_store.get()
+    status = dataclasses.replace(status, state="pause", volume=50, etag="changed")
+    player_mocks.player_data.status_store.set(status)
+
+    await asyncio.sleep(0)
+    for _ in range(10):
+        post_state = hass.states.get("media_player.player_name1111")
+        if post_state.state == MediaPlayerState.PAUSED:
+            break
+        await asyncio.sleep(1)
```

Comment:
We should not sleep 10 seconds in a test, if we added such logic to every integration we'd be waiting an hour more to run the full test suite :)

Reply:
The question is: How long should we wait? Too long and the test might block for to long when it is failing. Too short and we might have a flaky test. It is kind of difficult to test this.

Reply:
What should it wait for? We can freeze and skip time when needed. (I'm on mobile so I don't see the full context now)

Reply:
There is a loop waiting for `status(...)` to return a new value. I supply that value by resolving a Future. But i can not know when my value has been processed by the loop. Especially not if i am still waiting for the value to be processed or if processing failed and nothing is going to happen.

Reply:
Is there a possibility to patch the library to return such info?

Reply:
Not the library itself. The loop is within the `media_player` integration. I think it is impossible without looking into the internals of the integration.

Dropping the whole `ValueStore` thing and doing a busy loop, which runs every few milliseconds, might be the best solution. Assuming we do a loop every 10ms would mean we could be reasonably sure that we should have processed the changed state within 20ms.

Reply:
I removed the waiting(for up to 10 seconds). It works locally with `await hass.async_block_till_done()`.

This is only working because the test method is interrupted and the event loop has a chance to run. But i cannot guarantee that this will always be the case.